### PR TITLE
Fix/merge release3.6 to develop

### PR DESCRIPTION
**Pull request by sagarkapare**
Conflicts in 

```
AbstractContext.java
MapReduceProgramRunner.java
MapReduceRuntimeService.java
MapReduceTaskContextProvider.java
AbstractProgramTwillRunnable.java
FlowletProgramRunner.java
ServiceProgramRunner.java
WorkerProgramRunner.java
WorkflowPRogramRunner.java
ProgramLifeCycleService.java
DefaultUsageRegistry.java
DefaultSparkExecutionContext.java
tables.rst
UpgradeTool.java
SparkProgramRunner.java
DefaultSparkExecutionContext.scala
```
1. Most conflicts were because of the return type of Program.getId is changed in 3.6 to ProgramId from Id.Program.
2. Removal of cache in the DefaultUsageRegistry.
3. documentation changes in the tables.rst
4. Variable rename in UpgradeTool - datasetServiceManager is changed to upgradeDatasetServiceManager.

### CODE REVIEW

**Comment by sagarkapare** [comment]
Build is running here - https://builds.cask.co/browse/CDAP-DUT4882-1

File: cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java [hunk 1]
```diff
@@ -57,6 +58,7 @@
 /**
  * Tests for {@link WorkflowStatsSLAHttpHandler}
  */
+@Ignore
```
Comment (anwar6953) [inline cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java:61]: know why this happened? JIRA tracking it?

File: cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowHttpHandlerTest.java [hunk 2]
```diff
@@ -669,6 +669,7 @@ public Integer call() throws Exception {
 
   @Category(XSlowTests.class)
   @Test
+  @Ignore
```
Comment (anwar6953) [inline cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowHttpHandlerTest.java:672]: and these?

**Comment by anwar6953** [comment]
Also, can you merge release/3.6 into this branch (to get the latest changes that went into release/3.6).

File: cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java (hunk 1 above)
Comment (bdmogal) [reply cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java:61]: @maochf is working on a fix

File: cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowHttpHandlerTest.java (hunk 2 above)
Comment (bdmogal) [reply cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowHttpHandlerTest.java:672]: ditto

**Comment by sagarkapare** [comment]
`ProgramLifecycleHttpHandlerTest#testVersionedProgramStartStopStatus` is failing on this branch. looking further.

**Comment by sagarkapare** [comment]
Fixed test case and running the build again - https://builds.cask.co/browse/CDAP-DUT4882-3

**Comment by sagarkapare** [comment]
More conflicts in 

```
AbstractStreamFileConsumerFactory.java
DatasetSpecificationUpgrader.java
StreamStateStoreUpgrader.java
```

**Comment by anwar6953** [comment]
lgtm

**Comment by sagarkapare** [comment]
Build is green here - https://builds.cask.co/browse/CDAP-DUT4882-3. Merging.

**Pull request merged**